File: typescript/tests/debug/net-event-converter.test.ts
import { describe, it, expect } from 'vitest';
import { toEventInfo, tokenInfo, compactTokenInfo, convertMarking } from '../../src/debug/net-event-converter.js';
import { tokenOf, unitToken } from '../../src/core/token.js';
import type { NetEvent } from '../../src/event/net-event.js';
import type { TokenInfo } from '../../src/debug/debug-response.js';

const NOW = Date.now();

describe('NetEventConverter', () => {
  describe('toEventInfo', () => {
    it('should convert lifecycle events', () => {
      const started: NetEvent = {
        type: 'execution-started',
        timestamp: NOW,
        netName: 'MyNet',
        executionId: 'exec-1',
      };
      const info = toEventInfo(started);

      expect(info.type).toBe('ExecutionStarted');
      expect(info.transitionName).toBeNull();
      expect(info.placeName).toBeNull();
      expect(info.details['netName']).toBe('MyNet');
      expect(info.details['executionId']).toBe('exec-1');

      const completed: NetEvent = {
        type: 'execution-completed',
        timestamp: NOW,
        netName: 'MyNet',
        executionId: 'exec-1',
        totalDurationMs: 500,
      };
      const completedInfo = toEventInfo(completed);
      expect(completedInfo.type).toBe('ExecutionCompleted');
      expect(completedInfo.details['totalDurationMs']).toBe(500);
    });

    it('should convert transition events', () => {
      const enabled: NetEvent = {
        type: 'transition-enabled',
        timestamp: NOW,
        transitionName: 'T1',
      };
      const info = toEventInfo(enabled);
      expect(info.type).toBe('TransitionEnabled');
      expect(info.transitionName).toBe('T1');
      expect(info.placeName).toBeNull();

      const clockRestarted: NetEvent = {
        type: 'transition-clock-restarted',
        timestamp: NOW,
        transitionName: 'T2',
      };
      expect(toEventInfo(clockRestarted).type).toBe('TransitionClockRestarted');

      const started: NetEvent = {
        type: 'transition-started',
        timestamp: NOW,
        transitionName: 'T1',
        consumedTokens: [tokenOf('val')],
      };
      const startedInfo = toEventInfo(started);
      expect(startedInfo.type).toBe('TransitionStarted');
      expect(Array.isArray(startedInfo.details['consumedTokens'])).toBe(true);

      const tcompleted: NetEvent = {
        type: 'transition-completed',
        timestamp: NOW,
        transitionName: 'T1',
        producedTokens: [tokenOf('out')],
        durationMs: 42,
      };
      const completedInfo = toEventInfo(tcompleted);
      expect(completedInfo.type).toBe('TransitionCompleted');
      expect(completedInfo.details['durationMs']).toBe(42);

      const failed: NetEvent = {
        type: 'transition-failed',
        timestamp: NOW,
        transitionName: 'T1',
        errorMessage: 'boom',
        exceptionType: 'RuntimeException',
      };
      const failedInfo = toEventInfo(failed);
      expect(failedInfo.type).toBe('TransitionFailed');
      expect(failedInfo.details['errorMessage']).toBe('boom');
      expect(failedInfo.details['exceptionType']).toBe('RuntimeException');
    });

    it('should convert timeout events', () => {
      const transTimeout: NetEvent = {
        type: 'transition-timed-out',
        timestamp: NOW,
        transitionName: 'T1',
        deadlineMs: 1000,
        actualDurationMs: 1200,
      };
      const info = toEventInfo(transTimeout);
      expect(info.type).toBe('TransitionTimedOut');
      expect(info.transitionName).toBe('T1');
      expect(info.details['deadlineMs']).toBe(1000);
      expect(info.details['actualDurationMs']).toBe(1200);

      const actionTimeout: NetEvent = {
        type: 'action-timed-out',
        timestamp: NOW,
        transitionName: 'T2',
        timeoutMs: 500,
      };
      const actionInfo = toEventInfo(actionTimeout);
      expect(actionInfo.type).toBe('ActionTimedOut');
      expect(actionInfo.details['timeoutMs']).toBe(500);
    });

    it('should convert token events', () => {
      const token = tokenOf('hello');
      const added: NetEvent = {
        type: 'token-added',
        timestamp: NOW,
        placeName: 'P1',
        token,
      };
      const info = toEventInfo(added);
      expect(info.type).toBe('TokenAdded');
      expect(info.transitionName).toBeNull();
      expect(info.placeName).toBe('P1');
      expect(info.details['token']).toBeDefined();
      expect((info.details['token'] as TokenInfo).type).toBe('string');

      const removed: NetEvent = {
        type: 'token-removed',
        timestamp: NOW,
        placeName: 'P1',
        token,
      };
      const removedInfo = toEventInfo(removed);
      expect(removedInfo.type).toBe('TokenRemoved');
      expect(removedInfo.placeName).toBe('P1');
    });

    it('should convert marking snapshot', () => {
      const token = tokenOf('data');
      const snapshot: NetEvent = {
        type: 'marking-snapshot',
        timestamp: NOW,
        marking: new Map([['P1', [token]]]),
      };
      const info = toEventInfo(snapshot);

      expect(info.type).toBe('MarkingSnapshot');
      expect(info.transitionName).toBeNull();
      expect(info.placeName).toBeNull();
      expect(info.details['marking']).toBeDefined();
    });

    it('should convert log message with and without throwable', () => {
      const withoutThrowable: NetEvent = {
        type: 'log-message',
        timestamp: NOW,
        transitionName: 'T1',
        logger: 'com.example.Foo',
        level: 'INFO',
        message: 'hello',
        error: null,
        errorMessage: null,
      };
      const info = toEventInfo(withoutThrowable);
      expect(info.type).toBe('LogMessage');
      expect(info.transitionName).toBe('T1');
      expect(info.details['message']).toBe('hello');
      expect(info.details['level']).toBe('INFO');
      expect(info.details['throwable']).toBeUndefined();
      expect(info.details['throwableMessage']).toBeUndefined();

      const withThrowable: NetEvent = {
        type: 'log-message',
        timestamp: NOW,
        transitionName: 'T1',
        logger: 'com.example.Foo',
        level: 'ERROR',
        message: 'oops',
        error: 'RuntimeException',
        errorMessage: 'boom',
      };
      const throwInfo = toEventInfo(withThrowable);
      expect(throwInfo.details['throwable']).toBe('RuntimeException');
      expect(throwInfo.details['throwableMessage']).toBe('boom');
    });
  });

  describe('compact mode', () => {
    it('toEventInfo compact should omit token values', () => {
      const token = tokenOf('hello');
      const added: NetEvent = {
        type: 'token-added',
        timestamp: NOW,
        placeName: 'P1',
        token,
      };
      const info = toEventInfo(added, true);

      expect(info.type).toBe('TokenAdded');
      expect(info.placeName).toBe('P1');
      const tokenDetail = info.details['token'] as TokenInfo;
      expect(tokenDetail.type).toBe('string');
      expect(tokenDetail.value).toBeNull();
    });

    it('convertMarking compact should omit token values', () => {
      const token = tokenOf('data');
      const marking = new Map([['P1', [token]]]);
      const result = convertMarking(marking, true);

      expect(Object.keys(result)).toHaveLength(1);
      const tokens = result['P1']!;
      expect(tokens).toHaveLength(1);
      expect(tokens[0]!.type).toBe('string');
      expect(tokens[0]!.value).toBeNull();
    });

    it('convertMarking full should include token values', () => {
      const token = tokenOf('data');
      const marking = new Map([['P1', [token]]]);
      const result = convertMarking(marking, false);

      const tokens = result['P1']!;
      expect(tokens[0]!.type).toBe('string');
      expect(tokens[0]!.value).toBe('data');
    });
  });

  describe('tokenInfo conversion', () => {
    it('should return full value', () => {
      const longValue = 'x'.repeat(200);
      const token = tokenOf(longValue);
      const info = tokenInfo(token);
      expect(info.type).toBe('string');
      expect(info.value).toBe(longValue);
    });

    it('should handle unit token', () => {
      const unit = unitToken();
      const info = tokenInfo(unit);
      expect(info.type).toBe('null');
      expect(info.value).toBe('null');
    });

    it('compactTokenInfo should return type only', () => {
      const token = tokenOf('hello');
      const info = compactTokenInfo(token);
      expect(info.type).toBe('string');
      expect(info.value).toBeNull();
      expect(info.timestamp).toBeDefined();
    });
  });
});

File: java/src/test/java/org/libpetri/debug/NetEventConverterTest.java
package org.libpetri.debug;

import org.libpetri.core.Token;
import org.libpetri.debug.DebugResponse.NetEventInfo;
import org.libpetri.debug.DebugResponse.TokenInfo;
import org.libpetri.event.NetEvent;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class NetEventConverterTest {

    private static final Instant NOW = Instant.parse("2024-01-15T10:30:00Z");

    @Nested
    class ToEventInfo {

        @Test
        void shouldConvertLifecycleEvents() {
            var started = new NetEvent.ExecutionStarted(NOW, "MyNet", "exec-1");
            var info = NetEventConverter.toEventInfo(started);

            assertEquals("ExecutionStarted", info.type());
            assertEquals(NOW.toString(), info.timestamp());
            assertNull(info.transitionName());
            assertNull(info.placeName());
            assertEquals("MyNet", info.details().get("netName"));
            assertEquals("exec-1", info.details().get("executionId"));

            var completed = new NetEvent.ExecutionCompleted(NOW, "MyNet", "exec-1", Duration.ofMillis(500));
            var completedInfo = NetEventConverter.toEventInfo(completed);
            assertEquals("ExecutionCompleted", completedInfo.type());
            assertEquals(500L, completedInfo.details().get("totalDurationMs"));
        }

        @Test
        void shouldConvertTransitionEvents() {
            var enabled = new NetEvent.TransitionEnabled(NOW, "T1");
            var info = NetEventConverter.toEventInfo(enabled);
            assertEquals("TransitionEnabled", info.type());
            assertEquals("T1", info.transitionName());
            assertNull(info.placeName());

            var clockRestarted = new NetEvent.TransitionClockRestarted(NOW, "T2");
            assertEquals("TransitionClockRestarted", NetEventConverter.toEventInfo(clockRestarted).type());

            var started = new NetEvent.TransitionStarted(NOW, "T1", List.of(Token.of("val")));
            var startedInfo = NetEventConverter.toEventInfo(started);
            assertEquals("TransitionStarted", startedInfo.type());
            assertInstanceOf(List.class, startedInfo.details().get("consumedTokens"));

            var completed = new NetEvent.TransitionCompleted(NOW, "T1", List.of(Token.of("out")), Duration.ofMillis(42));
            var completedInfo = NetEventConverter.toEventInfo(completed);
            assertEquals("TransitionCompleted", completedInfo.type());
            assertEquals(42L, completedInfo.details().get("durationMs"));

            var failed = new NetEvent.TransitionFailed(NOW, "T1", "boom", "java.lang.RuntimeException");
            var failedInfo = NetEventConverter.toEventInfo(failed);
            assertEquals("TransitionFailed", failedInfo.type());
            assertEquals("boom", failedInfo.details().get("errorMessage"));
            assertEquals("java.lang.RuntimeException", failedInfo.details().get("exceptionType"));
        }

        @Test
        void shouldConvertTimeoutEvents() {
            var transitionTimeout = new NetEvent.TransitionTimedOut(NOW, "T1", Duration.ofMillis(1000), Duration.ofMillis(1200));
            var info = NetEventConverter.toEventInfo(transitionTimeout);
            assertEquals("TransitionTimedOut", info.type());
            assertEquals("T1", info.transitionName());
            assertEquals(1000L, info.details().get("deadlineMs"));
            assertEquals(1200L, info.details().get("actualDurationMs"));

            var actionTimeout = new NetEvent.ActionTimedOut(NOW, "T2", Duration.ofMillis(500));
            var actionInfo = NetEventConverter.toEventInfo(actionTimeout);
            assertEquals("ActionTimedOut", actionInfo.type());
            assertEquals(500L, actionInfo.details().get("timeoutMs"));
        }

        @Test
        void shouldConvertTokenEvents() {
            var token = Token.of("hello");
            var added = new NetEvent.TokenAdded(NOW, "P1", token);
            var info = NetEventConverter.toEventInfo(added);
            assertEquals("TokenAdded", info.type());
            assertNull(info.transitionName());
            assertEquals("P1", info.placeName());
            assertInstanceOf(TokenInfo.class, info.details().get("token"));

            var removed = new NetEvent.TokenRemoved(NOW, "P1", token);
            var removedInfo = NetEventConverter.toEventInfo(removed);
            assertEquals("TokenRemoved", removedInfo.type());
            assertEquals("P1", removedInfo.placeName());
        }

        @Test
        void shouldConvertMarkingSnapshot() {
            var token = Token.of("data");
            var marking = Map.<String, List<Token<?>>>of("P1", List.of(token));
            var snapshot = new NetEvent.MarkingSnapshot(NOW, marking);
            var info = NetEventConverter.toEventInfo(snapshot);

            assertEquals("MarkingSnapshot", info.type());
            assertNull(info.transitionName());
            assertNull(info.placeName());
            assertInstanceOf(Map.class, info.details().get("marking"));
        }

        @Test
        void shouldConvertLogMessageWithAndWithoutThrowable() {
            var withoutThrowable = new NetEvent.LogMessage(NOW, "T1", "com.example.Foo", "INFO", "hello", null, null);
            var info = NetEventConverter.toEventInfo(withoutThrowable);
            assertEquals("LogMessage", info.type());
            assertEquals("T1", info.transitionName());
            assertEquals("hello", info.details().get("message"));
            assertEquals("INFO", info.details().get("level"));
            assertFalse(info.details().containsKey("throwable"));
            assertFalse(info.details().containsKey("throwableMessage"));

            var withThrowable = new NetEvent.LogMessage(NOW, "T1", "com.example.Foo", "ERROR", "oops",
                "java.lang.RuntimeException", "boom");
            var throwInfo = NetEventConverter.toEventInfo(withThrowable);
            assertEquals("java.lang.RuntimeException", throwInfo.details().get("throwable"));
            assertEquals("boom", throwInfo.details().get("throwableMessage"));
        }
    }

    @Nested
    class CompactMode {

        @Test
        void toEventInfoCompactShouldOmitTokenValues() {
            var token = Token.of("hello");
            var added = new NetEvent.TokenAdded(NOW, "P1", token);
            var info = NetEventConverter.toEventInfo(added, true);

            assertEquals("TokenAdded", info.type());
            assertEquals("P1", info.placeName());
            var tokenInfo = (TokenInfo) info.details().get("token");
            assertEquals("String", tokenInfo.type());
            assertNull(tokenInfo.value());
        }

        @Test
        void convertMarkingCompactShouldOmitTokenValues() {
            var token = Token.of("data");
            var marking = Map.<String, List<Token<?>>>of("P1", List.of(token));
            var result = NetEventConverter.convertMarking(marking, true);

            assertEquals(1, result.size());
            var tokens = result.get("P1");
            assertEquals(1, tokens.size());
            assertEquals("String", tokens.getFirst().type());
            assertNull(tokens.getFirst().value());
        }

        @Test
        void convertMarkingFullShouldIncludeTokenValues() {
            var token = Token.of("data");
            var marking = Map.<String, List<Token<?>>>of("P1", List.of(token));
            var result = NetEventConverter.convertMarking(marking, false);

            var tokens = result.get("P1");
            assertEquals("String", tokens.getFirst().type());
            assertEquals("data", tokens.getFirst().value());
        }
    }

    @Nested
    class TokenInfoConversion {

        @Test
        void shouldReturnFullValueWithoutPreview() {
            var longValue = "x".repeat(200);
            var token = Token.of(longValue);
            var info = NetEventConverter.tokenInfo(token);
            assertEquals("String", info.type());
            assertEquals(longValue, info.value());
        }

        @Test
        void shouldHandleUnitToken() {
            var unit = Token.unit();
            var info = NetEventConverter.tokenInfo(unit);
            assertEquals("null", info.type());
            assertEquals("null", info.value());
        }

        @Test
        void compactTokenInfoShouldReturnTypeOnly() {
            var token = Token.of("hello");
            var info = NetEventConverter.compactTokenInfo(token);
            assertEquals("String", info.type());
            assertNull(info.value());
            assertNotNull(info.timestamp());
        }

        @Test
        void shouldProjectRecordIntoStructuredField() {
            var msg = new SampleMessage("USER", "hi", 2);
            var info = NetEventConverter.tokenInfo(Token.of(msg));

            assertEquals("SampleMessage", info.type());
            // toString stays as the human-friendly display form
            assertTrue(info.value().contains("SampleMessage"));
            // Structured field is a JsonNode-like projection — consumers read named components
            assertNotNull(info.structured(), "records with public accessors must project");
            var structured = info.structured().toString();
            assertTrue(structured.contains("USER"));
            assertTrue(structured.contains("hi"));
        }

        @Test
        void shouldEmitEnumNameInStructuredField() {
            var info = NetEventConverter.tokenInfo(Token.of(SampleMode.VOICE));

            assertEquals("SampleMode", info.type());
            assertEquals("VOICE", info.structured(),
                "enums surface as the name() string so LLM responses stay compact");
        }

        @Test
        void shouldPassPrimitivesThroughStructured() {
            assertEquals(42, NetEventConverter.tokenInfo(Token.of(42)).structured());
            assertEquals("hello", NetEventConverter.tokenInfo(Token.of("hello")).structured());
            assertEquals(true, NetEventConverter.tokenInfo(Token.of(true)).structured());
        }

        @Test
        void shouldNotPopulateStructuredForUnitTokens() {
            var info = NetEventConverter.tokenInfo(Token.unit());
            assertNull(info.structured(),
                "unit tokens have no meaningful payload — structured stays null");
        }

        @Test
        void shouldLeaveStructuredNullForOpaqueBeans() {
            var info = NetEventConverter.tokenInfo(Token.of(new OpaqueBean()));

            assertEquals("OpaqueBean", info.type());
            assertNull(info.structured(),
                "Jackson-empty beans must not inflate the response with an empty object");
        }
    }

    public record SampleMessage(String kind, String text, int length) {}

    public enum SampleMode { TEXT, VOICE }

    /** Has no public getters and no Jackson annotations → Jackson produces an empty node. */
    public static final class OpaqueBean {
        @SuppressWarnings("unused")
        private final String secret = "hidden";
        @Override public String toString() { return "OpaqueBean{***}"; }
    }
}
